domain/agentbinary/types_test.go

```diff
@@ -0,0 +1,269 @@
+func (_ *typesSuite) TestAgentBinaryHighestVersion(c *tc.C) {
+	version1, err := semversion.Parse("4.0.1")
+	c.Assert(err, tc.ErrorIsNil)
+	version2, err := semversion.Parse("4.0-beta1")
+	c.Assert(err, tc.ErrorIsNil)
+	version3, err := semversion.Parse("4.1.0")
+	c.Assert(err, tc.ErrorIsNil)
+
+	agentBinaries := []AgentBinary{
+		{
+			Architecture: AMD64,
+			Stream:       AgentStreamReleased,
+			Version:      version3,
+		},
+		{
+			Architecture: AMD64,
+			Stream:       AgentStreamReleased,
+			Version:      version2,
+		},
+		{
+			Architecture: AMD64,
+			Stream:       AgentStreamReleased,
+			Version:      version1,
+		},
+	}
+
+	ver := slices.MaxFunc(agentBinaries, AgentBinaryHighestVersion)
+	mc := tc.NewMultiChecker()
+	mc.AddExpr("_[_].Architecture", tc.Ignore)
+	mc.AddExpr("_[_].Stream", tc.Ignore)
+	c.Check(ver, mc, AgentBinary{Version: version3})
+}
+
+// TestAgentBinaryNotMatchinVersion tests the [AgentBinaryNotMatchingVersion]
+// with [slices.Delete] to confirm that it removes all [AgentBinaries] that do
+// match a target version.
+func (*typesSuite) TestAgentBinaryNotMatchinVersion(c *tc.C) {
+	versionToMatch, err := semversion.Parse("4.0.1")
+	c.Assert(err, tc.ErrorIsNil)
+
+	version1 := versionToMatch
+	version2, err := semversion.Parse("4.0-beta1")
+	c.Assert(err, tc.ErrorIsNil)
+	version3, err := semversion.Parse("4.1.0")
+	c.Assert(err, tc.ErrorIsNil)
+
+	agentBinaries := []AgentBinary{
+		{
+			Architecture: AMD64,
+			Stream:       AgentStreamReleased,
+			Version:      version1,
+		},
+		{
+			Architecture: ARM64,
+			Stream:       AgentStreamProposed,
+			Version:      version1,
+		},
+		{
+			Architecture: S390X,
+			Stream:       AgentStreamReleased,
+			Version:      version2,
+		},
+		{
+			Architecture: S390X,
+			Stream:       AgentStreamReleased,
+			Version:      version3,
+		},
+		{
+			Architecture: ARM64,
+			Stream:       AgentStreamReleased,
+			Version:      version3,
+		},
+	}
+
+	agentBinaries = slices.DeleteFunc(
+		agentBinaries, AgentBinaryNotMatchingVersion(versionToMatch),
+	)
+	c.Check(agentBinaries, tc.SameContents, []AgentBinary{
+		{
+			Architecture: AMD64,
+			Stream:       AgentStreamReleased,
+			Version:      version1,
+		},
+		{
+			Architecture: ARM64,
+			Stream:       AgentStreamProposed,
+			Version:      version1,
+		},
+	})
+}
+
+// TestArchitectureFromStringNotRecognised ensures that if the string
+// architecture is not recognised then the returned Architecture is 0 with a
+// false conversion value.
+func (_ *typesSuite) TestArchitectureFromStringNotRecognised(c *tc.C) {
+	arch, converted := ArchitectureFromString("fakearch")
+	c.Check(converted, tc.IsFalse)
+	c.Check(arch, tc.Equals, Architecture(0))
+}
+
+// TestArchitectureFromString ensures that [ArchitectureFromString] correctly
+// converts known architecture strings to their corresponding [Architecture]
+// values.
+func (_ *typesSuite) TestArchitectureFromString(c *tc.C) {
```

Comment:
```suggestion
func (*typesSuite) TestArchitectureFromString(c *tc.C) {
```

nit: drop all of these to be consistent